### DISCUSSION TITLE
Can't load/use multiple samples loaded from SD card [solved]

### DISCUSSION THREAD

**Discussion by souravlal** [marcel-licence/esp32_midi_sampler#48] [Q&A]
With the SD card I can load and play sample from SD card but if I look to choose a different sample the previously selected sample plays, it  needs a restart of the ESP32 Audio-kit to load a new sample from SD card.

**Comment by marcel-licence** [comment]
Can you explain in detail what you did?

**Reply by souravlal** [reply]
I have connected the display and I loaded an 8gb sd card with wav samples. I am using an Arduino to send midi messages and I can select sd card as the storage device in the display. I can also load a sample from sd card and it plays according to the midi messages. But when I try to load another sample it does not load up and the  previous selected sample gets triggered according to the midi signal. I have to restart the ESP32 Audio-kit to load a new sample from sd card.

Same is the case with the recording sample,  I can record a sample successfully and play it, buf if I try to record again the display shows 'sample recorded' but the previously recorded sample gets triggered. If I restart the ESP32 Audio-kit I can record and play a new sample.

Also in page 'Effects' the 'delay feedback' option can't be accessed using buttons.Also the ADSR can't be adjusted by buttons.

**Reply by marcel-licence** [reply]
It may be not implemented. I've just prepared a little starting point which may help you for your own implementation.

**Reply by souravlal** [reply]
@marcel-licence So you are saying that to load a new sample I need to restart the ESP32 Audio-kit each time?

**Reply by marcel-licence** [reply]
A restart is not necessary to reload samples

**Reply by souravlal** [reply]
@marcel-licence Do you know what could be the problem then? Is there anything I need to change in the code? I have uploaded the photo of my program settings above.Is it the problem of SD Card?
MIDI Channel 2 I am using, is that ok?

**Reply by souravlal** [reply]
Hey guys do anyone faced a similar problem?👆
Can anyone please help me out?

**Comment by souravlal** [comment]
Hey guys do anyone faced a similar problem?👆
Can anyone please help me out?
Firstly I purchased an ESP32 Audio-kit with the ES8388 codec, since I realized that your code doesn't support it I purchased a ESP32A1S chip with AC101 codec and soldered it to my board by removing the chip with ES8388 codec. I want to know whether this could be the problem, or it just doesn't matter, all the other functions are working correctly.
Could it be a software problem or hardware problem?

This is my settings, are these correct?:-

![IMG20210907205542](https://user-images.githubusercontent.com/53053463/132371394-473dde25-58ae-43bb-88ad-f889c4f584f6.jpg)

**Reply by marcel-licence** [reply]
Did you keep in mind that new loaded samples are played on different midi channels?

**Reply by souravlal** [reply]
So do I have to change the MIDI channel each time I load a sample?
 Generally I was using MIDI channel 2, after Loading a different sample do I need to change the channel to 3?
In commonly used samplers we just need to load the samples, don't need to change the MIDI channel, right?
So after Loading the 17th sample, which channel should  I select? Because there is only 16 MIDI channels, right?
Forgive me if I am asking foolish questions.

**Reply by marcel-licence** [reply]
Yes. Samples are connected to different channels in this version. That is the reason you are starting with channel 2 because on 1 you would have one key for each sample.

There is also the note on top of the sampler_module.ino file:
` * MIDI ch1 noteOn message will trigger different samples
 * MIDI ch2-16 noteOn will trigger each a certain sample with different pitch`

Yes 16 channels is the limit.
To use more 'instruments' the code needs a modification. I was working on a little update coming soon to select an instrument using CC. That would allow to select an individual instrument per channel. 

When looking into 'Sampler_NoteOnInt' you will find a check 'if (ch == 0)' ... 'else' 

        'struct sample_record_s *rec = &sampleRecords[(ch - 1) % sampleRecordCount]; /* decrease by one because we want to start with the first sample here *'
This line is doing the sample selection. You can modify this line for your own needs.

Hope this helps

**Reply by souravlal** [reply]
@marcel-licence So after I load 16 samples which will be played on different 16 channels like you said, if I want to load a new one I mean 17th sample which channel it will get played? Or if I need to load a sample after the 16th one do I need to restart the Audio kit?
Won't it be possible to overwrite the old sample with the new sample we load and play it on a certain channel only(like MIDI channel 2), Isn't it more comfortable to use,right? Is it hard to implement in code? At the moment the 16 samples are saved on PSRAM simultaneously if I load them, right?

**Reply by marcel-licence** [reply]
A restart is not required. You can also call Sampler_RemoveActiveRecording (you will also find it in the midi mapping). 
Yes all samples are stored in the PSRAM, so the limit is 4MB. Otherwise you wouldn't get the performance as shown in my videos.

**Reply by souravlal** [reply]
Is it the same case with the 'sample record'?

Won't it be possible to overwrite the old sample with the new sample we load and play it on a certain channel only(like MIDI channel 2), Isn't it more comfortable to use,right? Is it hard to implement in code?

**Reply by marcel-licence** [reply]
It would be possible. I requires some more effort. The problem is the samples do not have always the same length. You have linear storage. So it might require to shift data or it may cause some fragmentation. There are a lot of possibilities. But all require some little time for an implementation.

**Reply by souravlal** [reply]
Ok
